apis/field_error.go

```diff
@@ -28,18 +28,55 @@ import (
 // a problem with the current field itself.
 const CurrentField = ""
 
+// DiagnosticLevel is used to signal the severity of a particular diagnostic
+// in the form of a FieldError.
+type DiagnosticLevel string
+
+const (
+	// ErrorLevel is used to signify fatal/blocking diagnostics, e.g. those
+	// that should block admission in a validating admission webhook.
+	ErrorLevel DiagnosticLevel = "Error"
+
+	// WarningLevel is used to signify information/non-blocking diagnostics,
+	// e.g. those that should be surfaced as warnings in a validating admission
+	// webhook.
+	WarningLevel DiagnosticLevel = "Warning"
+)
+
+// Matches checks whether the provided diagnostic level matches this one,
+// including the special case where an empty DiagnosticLevel equals ErrorLevel.
+func (l DiagnosticLevel) Matches(r DiagnosticLevel) bool {
+	switch {
+	case l == ErrorLevel && r == "":
+		return true
+	case r == ErrorLevel && l == "":
+		return true
+	default:
+		return l == r
+	}
+}
+
 // FieldError is used to propagate the context of errors pertaining to
 // specific fields in a manner suitable for use in a recursive walk, so
 // that errors contain the appropriate field context.
 // FieldError methods are non-mutating.
 // +k8s:deepcopy-gen=true
 type FieldError struct {
+	// Message holds the main diagnostic message carried by this FieldError
 	Message string
-	Paths   []string
+
+	// Paths holds a list of paths to which this diagnostic pertains
+	Paths []string
+
+	// Level holds the severity of the diagnostic.
+	// If empty, this defaults to ErrorLevel.
+	Level DiagnosticLevel
```

Comment:
Do we want `Level` to be a string or an int (using https://pkg.go.dev/golang.org/x/tools/cmd/stringer to convert to string when needed)?

Reply:
i.e.

```golang

type Level int

// go:generate stringer -type=Level
const (
  ErrorLevel Level = iota
  DiagnosticLevel
)

Reply:
Hmm, `iota` isn't the zero value (AFAIK), so I'm not sure this WAI with the back-compat goal...  Maybe I'm missing something? 🤔

Reply:
`iota` starts at the zero value at the beginning of each const block, and then increases for each subsequent ConstSpec (line) to the end of the block.